design/design-doc-custom-composition.md

```diff
@@ -0,0 +1,179 @@
+# Custom Composition Types
+
+* Owner: Muvaffak Onuş (@muvaf)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+## Proposal
+
+### Authorship Experience
+
+Let's say we have the following `CompositeResourceDefinition` that defines a `PrivateNetwork`
+type.
+
+```yaml
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: InfrastructureDefinition
+metadata:
+  name: privatenetworks.aws.example.org
+spec:
+  crdSpecTemplate:
+    group: aws.example.org
+    version: v1alpha1
+    names:
+      kind: PrivateNetwork
+      listKind: PrivateNetworkList
+      plural: privatenetworks
+      singular: privatenetwork
+    validation:
+      openAPIV3Schema:
+        type: object
+        properties:
+          spec:
+            type: object
+            properties:
+              region:
+                type: string
+                description: Geographic region in which nodes will exist.
+              providerConfigRef:
+                type: object
+                description: Crossplane AWS provider credentials to use.
+                properties:
+                  name:
+                    type: string
+                required:
+                - name
+            required:
+            - region
+            - providerConfigRef
+```
+
+User can have multiple `Composition` or custom compositions that target this new
+type. Let's assume they would like to write a cdk8s app and use it as the composition
+engine. They'd need to write the cdk8s app that will take the following resources
+as input:
+  * `PrivateNetwork` custom resource.
+  * `CustomComposition` custom resource.
+  It will return all the resulting resources as output.
+
+This conversion function will be wrapped by an HTTP server and pushed as an image
+whose `ENTRYPOINT` is to start that HTTP server. The input will come as POST
+request and the body of the result will consist of all the output resources.
```

Comment:
Would this pattern work for other 'custom' tools besides cdk8s? Presumably we could (for example) write a Docker image that invoked Kustomize or similar via a compatible HTTP API.

Reply:
> Presumably we could (for example) write a Docker image that invoked Kustomize or similar via a compatible HTTP API.

Yeah, that's what I had in mind. Something need to be the wrapper server and give the composite CR as input to these tools.